### ISSUE TITLE
Index gets rebuilt for each file in repo

### ISSUE THREAD

**Issue by mjomble** [isomorphic-git/isomorphic-git#1734]
Steps to reproduce:

* Create a repo with two or more files in the latest commit
* Stage changes in one file
* Run the following script in Node.js 18:

```javascript
const fs = require('fs')
const git = require('isomorphic-git') // version 1.21.0
const dir = TODO // Point this to the repo

git.walk({
  fs: {
    ...fs,
    promises: {
      ...fs.promises,
      writeFile: (file, data, options) => {
        console.log('updating', file)
        return fs.promises.writeFile(file, data, options)
      },
    },
  },
  dir,
  trees: [git.WORKDIR()],
  map: async (_, [workdir]) => workdir?.oid(),
})
.then(() => console.log('done'), console.error)
```

Expected outcome:

The index file is not modified (no "updating" lines are logged)

Actual outcome:

The index file is rebuilt once per each file in repo and ends up different from the one generated by Git (version 2.31.1.windows.1)
Running the script again after that results in no index modifications. But after unstaging and re-staging, the script rebuilds the index again on the next run.

If you only unstage the file via git between two runs, this script causes the index to be rebuilt only once. Although even in this case, it seems like it shouldn't be rebuilt at all.

----

The above code is a simplified reproduce scenario where the rebuild has no performance impact.
But the codebase where I initially discovered the problem has 1700+ files and a 300+ KB index file.
It took 40-50 seconds to rebuild it so many times. But only 1-2 seconds to walk the repo after the rebuild runs.

**Comment by mjomble** [comment]
After deeper investigation, I found that the root cause was that the Windows equivalents for the `ino` and `ctime` stat fields (which are written into the index file) are ambiguous.

Node.js tries to approximate `ino` and `dev` using alternative features, whereas the Windows version of Git just uses 0 for both.
And even though `ctimeMs` seems like the best fit for `ctime`, in the Windows version of Node.js the `birthtimeMs` field matches the value used by Git.

In other words, it seems isomorphic-git should continue to use `stat.ctimeMs`, `stat.dev` and `stat.ino` on Unix, but on Windows, it should use `stat.birthtimeMs`, `0` and `0`.

I can submit a PR if needed, but don't mind if anyone else beats me to it 😄

**Comment by jcubic** [comment]
If you can provide a fix that would be great make sure to create unit tests that fail without a fix if possible.

**Comment by mjomble** [comment]
I'll look into that.

In the meantime, here's a temporary workaround in case anyone needs it:

```javascript
const os = require('os')
const isWindows = os.platform() === 'win32'

git.walk({
  fs: {
    ...fs,
    promises: {
      ...fs.promises,
      lstat: async (path, opts) => {
        const stat = await fs.promises.lstat(path, opts)

        if (isWindows) {
          stat.ctimeMs = stat.birthtimeMs
          stat.ino = 0
          stat.dev = 0
        }

        return stat
      },
    },
  },
  // Rest of the parameters
})
```
